File: src/main/java/com/ocunha/domain/restaurant/filters/FilterDefinition.java
package com.ocunha.domain.restaurant.filters;

import com.ocunha.domain.restaurant.model.RestaurantSearchParams;
import com.ocunha.domain.restaurant.model.model.Restaurant;
import org.apache.commons.lang3.StringUtils;

import static java.util.Objects.isNull;

@FunctionalInterface
public interface FilterDefinition {

    boolean filter(Restaurant restaurant, RestaurantSearchParams searchParams);

    default boolean isGreaterOrEquals(Integer current, Integer searchParam) {
        return isNull(searchParam) || current.compareTo(searchParam) >= 0;
    }

    default boolean isLessOrEquals(Integer current, Integer searchParam) {
        return isNull(searchParam) || current.compareTo(searchParam) <= 0;
    }

    default boolean contains(String current, String searchParam) {
        return isNull(searchParam) || StringUtils.containsIgnoreCase(current, searchParam);
    }

}

File: src/main/kotlin/com/ocunha/domain/filter/FilterDefinition.kt
package com.ocunha.domain.filter

import com.ocunha.domain.restaurant.model.Restaurant
import com.ocunha.domain.restaurant.model.RestaurantSearchParams
import org.apache.commons.lang3.StringUtils
import java.util.Objects.isNull


interface FilterDefinition {

    fun filter(restaurant: Restaurant?, searchParams: RestaurantSearchParams?): Boolean

    fun isGreaterOrEquals(current: Int?, searchParam: Int?): Boolean {
        return isNull(searchParam) || current!! >= searchParam!!
    }

    fun isLessOrEquals(current: Int?, searchParam: Int?): Boolean {
        return isNull(searchParam) || current!! <= searchParam!!
    }

    fun contains(current: String?, searchParam: String?): Boolean {
        return isNull(searchParam) || StringUtils.containsIgnoreCase(current, searchParam)
    }
}